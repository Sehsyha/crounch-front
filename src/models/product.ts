export interface ProductResponse {
  data: Products;
}

export interface Products {
  products: ProductApi[];
}

export interface ProductApi {
  product_name: string;
  categories: string;
  code: string;
}

export interface Product {
  barCode: string;
  name: string;
  categories: string[];
}
