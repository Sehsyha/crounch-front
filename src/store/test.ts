import { ActionTree, MutationTree, GetterTree, Module, ActionContext } from 'vuex';
import { RootState } from '.';

export type ActionCall<S, R> = (injectee: ActionContext<S, R>, payload?: any) => any;

export function createStoreModuleMock<S, A, M, G>(initialState: S, actionKeys: A, mutationKeys: M, getterKeys: G): Module<S, RootState> {
  const actions: ActionTree<S, RootState> = {};
  const mutations: MutationTree<S> = {};
  const getters: GetterTree<S, RootState> = {};

  Object.keys(actionKeys).forEach(key => actions[key] = jest.fn());
  Object.keys(mutationKeys).forEach(key => mutations[key] = jest.fn());
  Object.keys(getterKeys).forEach(key => getters[key] = jest.fn());

  return {
    namespaced: true,
    actions,
    getters,
    mutations,
    state: initialState,
  }
}
