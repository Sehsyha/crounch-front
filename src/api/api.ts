import { getAPIURL } from '@/utils/environment';
import { TOKEN_STORAGE_KEY } from '@/utils/constants';
import { FetchError } from '@/utils/error';
import { AuthModule } from '@/store/AuthModule';
import { getModule } from 'vuex-module-decorators';
import router from '@/router/router';

export interface FetchOptions {
  url: string;
  data?: object;
  method: string;
}

export function doFetch<T>(options: FetchOptions): Promise<T> {
  return fetch(getUrl(options.url), getFetchOptions(options))
    .then(res => {
      if (res.status >= 200 && res.status <= 299 && res.status !== 204) {
        return res.json()
      }
      if (res.status === 204) {
        return
      }
      if (res.status === 401) {
        const authModule = getModule(AuthModule)
        authModule.logoutAction().then(() => {
          router.push({ name: 'home' })
        })
      }

      return res.json()
        .then(body => { throw new FetchError('Erreur lors de la requÃªte', res.status, body) })
        .catch()
    })
}

function getFetchOptions({ data, method }: FetchOptions): object {
  let options: any = {
    method,
    headers: {
      Authorization: localStorage.getItem(TOKEN_STORAGE_KEY),
    },
  };

  if (method !== 'GET') {
    options.body = JSON.stringify(data);
  }

  return options;
}

function getUrl(url: string): string {
  return `${getAPIURL()}/${url}`;
}
