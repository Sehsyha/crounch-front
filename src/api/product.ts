import { doFetch } from './api';
import { Product } from '@/models/product';

export function createProduct(name: string): Promise<Product> {
  return doFetch({
    url: 'products',
    method: 'POST',
    data: { name }
  });
}

export function searchProduct(name: string): Promise<Array<Product>> {
  return doFetch({
    url: 'products/search',
    method: 'POST',
    data: { name }
  })
}
