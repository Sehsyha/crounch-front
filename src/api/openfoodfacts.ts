import axios from 'axios';
import { ProductApi } from '@/models/product';

export function search(name: string): Promise<ProductApi[]> {
  return axios
    .get(`https://fr.openfoodfacts.org/cgi/search.pl?action=process&tagtype_0=categories&tag_contains_0=contains&tag_0=${name}&json=true`)
    .then(res => res.data);
}
