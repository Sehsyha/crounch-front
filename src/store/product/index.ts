import { Module } from 'vuex';

import { actions } from './actions';
import { getters } from './getters';
import { mutations } from './mutations';
import { RootState } from '..';
import { Product } from '@/models/product';

export interface ProductState {
  products: Product[];
}

export const state: ProductState = {
  products: [],
};

const namespaced: boolean = true;
export const productNamespace = { namespace: 'product' };

export const product: Module<ProductState, RootState> = {
  namespaced,
  state,
  actions,
  getters,
  mutations,
};
