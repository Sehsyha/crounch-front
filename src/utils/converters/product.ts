import { ProductApi, Product } from '@/models/product';

export function convertProductFromApi(productsApi: ProductApi[]): Product[] {
  return productsApi.map(productApi => ({
    name: productApi.product_name,
    categories: productApi.categories.replace(' ', '').split(','),
    barCode: productApi.code,
  }));
}
