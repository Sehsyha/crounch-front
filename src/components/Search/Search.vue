<template>
  <div>
    <form>
      <input type="text" v-model="name" @keyup="search" placeholder="Nom du produit">
    </form>
    <div id="products" v-if="this.products.length">
      <div class=".products" v-for="product in this.products" :key="product.code">
        {{ product.name }}
        {{ product.barCode }}
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Component } from 'vue-property-decorator';

import { search } from '@/api/openfoodfacts';
import { Product } from '@/models/product';
import { convertProductFromApi } from '@/utils/converters/product';

@Component
export default class Search extends Vue {
  name: string = '';
  products: Product[] = [];

  search() {
    search(this.name).then(products => {
      this.products = convertProductFromApi(products);
    });
  }
}
</script>
