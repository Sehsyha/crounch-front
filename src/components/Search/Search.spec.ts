import { Wrapper } from '@vue/test-utils';

import Search from './Search.vue';
import { shallowComponent } from '@/utils/test';
import { search } from '@/api/openfoodfacts';
import { convertProductFromApi } from '@/utils/converters/product';
import { ProductApi, Product } from '@/models/product';
import { when } from 'jest-when';

jest.mock('@/api/openfoodfacts');
jest.mock('@/utils/converters/product');

describe('Search', () => {
  let wrapper: Wrapper<Search>;
  const name = 'Saucisse';

  const productsApi: ProductApi[] = [
    {
      product_name: 'product name',
      code: '82793',
      categories: 'Hello, World',
    },
  ];

  const products: Product[] = [
    {
      name: 'product name',
      barCode: '82793',
      categories: ['Hello', 'World'],
    },
  ];

  beforeEach(() => {
    when(search as jest.Mock)
      .calledWith(name)
      .mockResolvedValue(productsApi);
    when(convertProductFromApi as jest.Mock)
      .calledWith(productsApi)
      .mockReturnValue(products);
    wrapper = shallowComponent(Search);
  });

  it('Should render.', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('Init', () => {
    it('Should initialize name with empty string.', () => {
      expect(wrapper.vm.$data.name).toBe('');
    });

    it('Should initialize products with an empty array.', () => {
      expect(wrapper.vm.$data.products).toEqual([]);
    });
  });

  it('Should call api when key is up in input.', () => {
    searchProduct();

    expect(search).toBeCalledWith(name);
  });

  it('Should convert products when the request succeed', () => {
    searchProduct();

    expect(convertProductFromApi).toHaveBeenCalledWith(productsApi);
  });

  it('Should set products from the request when it succeed.', done => {
    searchProduct();

    setTimeout(() => {
      expect(wrapper.vm.$data.products).toEqual(products);
      done();
    });
  });

  it('Should display the products when the request succeed.', done => {
    searchProduct();

    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  it('Should not display products when there is no products.', () => {
    expect(wrapper.find('#products').exists()).toBeFalsy();
  });

  function searchProduct() {
    wrapper.setData({ name });

    wrapper.find('input').trigger('keyup');
  }
});
