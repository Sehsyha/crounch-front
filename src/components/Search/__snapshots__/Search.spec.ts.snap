// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Should display the products when the request succeed. 1`] = `
<div>
  <form><input type="text" placeholder="Nom du produit"></form>
  <div id="products">
    <div class=".products">
      product name
      82793
    </div>
  </div>
</div>
`;

exports[`Search Should render. 1`] = `
<div>
  <form><input type="text" placeholder="Nom du produit"></form>
  <!---->
</div>
`;
